fusor-ember-cli/app/controllers/assign-nodes.js

```diff
@@ -98,50 +80,51 @@ export default Ember.Controller.extend(DeploymentControllerMixin, NeedsDeploymen
     }
   }),
 
-  showLoadingSpinner: false,
-  loadingSpinnerText: "Loading...",
-
   doAssignRole(plan, role, profile) {
-    var data;
-    var self = this;
+    var data, self = this,
+      unassignedRoles = this.get('unassignedRoles'),
+      assignedRoles = this.get('assignedRoles');
 
-    if (profile == null ) {
-      var unassignedRoles = this.get('unassignedRoles');
+    if (profile === null) {
       if (unassignedRoles.contains(role)) {
         // Role is already unassigned, do nothing
         return;
       }
-      data = { 'role_name': role.get('name'), 'flavor_name': null };
+      assignedRoles.removeObject(role);
+      data = {'role_name': role.get('name'), 'flavor_name': null};
     } else {
-      data = { 'role_name': role.get('name'), 'flavor_name': profile.get('name') };
+      unassignedRoles.removeObject(role);
+      data = {'role_name': role.get('name'), 'flavor_name': profile.get('name')};
     }
 
-    self.set('loadingSpinnerText', "Loading...");
-    self.set('showLoadingSpinner', true);
+    self.set('numRequests', self.get('numRequests') + 1);
     var token = Ember.$('meta[name="csrf-token"]').attr('content');
-    //ic-ajax request
-    console.log('PUT /fusor/api/openstack/deployments/' + this.get('deploymentId') + '/deployment_plans/overcloud/update_role_flavor');
+
     request({
-        url: '/fusor/api/openstack/deployments/' + this.get('deploymentId') + '/deployment_plans/overcloud/update_role_flavor',
-        type: 'PUT',
-        headers: {
-          "Accept": "application/json",
-          "Content-Type": "application/json",
-          "X-CSRF-Token": token,
-        },
-        data: JSON.stringify(data)
-      }).then(function(result) {
-        self.set('showLoadingSpinner', false);
-        console.log('SUCCESS');
-        self.store.push('deployment_plan', self.store.normalize('deployment_plan', result.deployment_plan));
-      }, function(error) {
-        error = error.jqXHR;
+      url: '/fusor/api/openstack/deployments/' + this.get('deploymentId') + '/deployment_plans/overcloud/update_role_flavor',
+      type: 'PUT',
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+        'X-CSRF-Token': token,
+      },
+      data: JSON.stringify(data)
+    }).then(function (result) {
+        self.set('numRequests', self.get('numRequests') - 1);
+        if (self.get('numRequests') <= 0) {
```

Comment:
I'm not calling refresh page since the request returns the updated plan, so it's just using that and refreshing the store.  The UI isn't blocked for any of these since I'm not using the loading spinner which takes the whole page.

For assign roles I wasn't sure about what server-side logic on the OSP side might be going on while doing so.  For that, once you drop the role, it will disappear to prevent the user from editing it until the request returns the updated deployment.  However, the user can go ahead and drop all the rest of the roles while this is going on.  When all requests are complete, the plan refreshes and the user has current data.  It seemed like a nice compromise to leave the UI usable but also refresh the data from the server in a non-disruptive way.

For the simple requests that called update_parameters and pushed the data since I was that was only a PUT of parameters.  I also did the same for role_count.